ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt

```diff
@@ -16,15 +16,116 @@
 package com.google.android.ground.ui.datacollection.tasks.multiplechoice
 
 import android.content.res.Resources
+import android.text.Editable
+import android.text.TextWatcher
+import com.google.android.ground.model.job.Job
+import com.google.android.ground.model.submission.MultipleChoiceResponse
 import com.google.android.ground.model.submission.MultipleChoiceResponse.Companion.fromList
+import com.google.android.ground.model.submission.Value
 import com.google.android.ground.model.task.Option
+import com.google.android.ground.model.task.Task
 import com.google.android.ground.ui.datacollection.tasks.AbstractTaskViewModel
 import javax.inject.Inject
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.filterNotNull
 
 class MultipleChoiceTaskViewModel @Inject constructor(resources: Resources) :
   AbstractTaskViewModel(resources) {
 
-  fun updateResponse(options: List<Option>) {
-    setValue(fromList(task.multipleChoice, options.map(Option::id)))
+  private val _items: MutableStateFlow<List<MultipleChoiceItem>> = MutableStateFlow(emptyList())
+  val itemsFlow: Flow<List<MultipleChoiceItem>> = _items.asStateFlow().filterNotNull()
+
+  private val selectedIds: MutableSet<String> = mutableSetOf()
+  private var otherText: String = ""
+
+  val textWatcher =
+    object : TextWatcher {
+      override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
+        otherText = s.toString()
+        updateResponse()
+      }
+
+      override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
+        // Not implemented.
+      }
+
+      override fun afterTextChanged(s: Editable) {
+        // Not implemented.
+      }
+    }
+
+  override fun initialize(job: Job, task: Task, value: Value?) {
+    super.initialize(job, task, value)
+    loadPendingSelections()
+    updateMultipleChoiceItems()
+  }
+
+  fun toggleItem(item: MultipleChoiceItem, canSelectMultiple: Boolean) {
+    val wasSelected = selectedIds.contains(item.option.id)
+    val isSelected = !wasSelected
+    if (!canSelectMultiple) {
+      selectedIds.clear()
+    }
+    if (isSelected) {
+      selectedIds.add(item.option.id)
+    } else {
+      selectedIds.remove(item.option.id)
+    }
+    updateResponse()
+    updateMultipleChoiceItems()
+  }
+
+  fun updateResponse() {
+    setValue(
+      fromList(
+        task.multipleChoice,
+        selectedIds.map {
+          // if the other option is selected grab the associated text save instead of otherId
+          if (it == OTHER_ID) OTHER_PREFIX + otherText.trim() + OTHER_SUFFIX else it
```

Comment:
This workaround sounds like something that would be specific to the remote persistence layer, not the view. Instead, I would expect the "other" text to be set in the `otherText` field we added to the model, and our ad hoc encoding only represented in one of the remote persistence converters . Can we do that as a fast-follow?

Reply:
Yeah makes sense, lets make a bug for it.